src/workerd/api/gpu/gpu-async-runner.h

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+// Based on the dawn node bindings
+
+#pragma once
+
+#include "workerd/io/io-context.h"
+#include <kj/timer.h>
+#include <webgpu/webgpu_cpp.h>
+
+namespace workerd::api::gpu {
+
+// AsyncRunner is used to poll a wgpu::Device with calls to Tick() while there
+// are asynchronous tasks in flight.
+class AsyncRunner {
+public:
+  AsyncRunner(wgpu::Device device) : device_(device){};
+
+  // Begin() should be called when a new asynchronous task is started.
+  // If the number of executing asynchronous tasks transitions from 0 to 1, then
+  // a function will be scheduled on the main JavaScript thread to call
+  // wgpu::Device::Tick() whenever the thread is idle. This will be repeatedly
+  // called until the number of executing asynchronous tasks reaches 0 again.
+  void Begin();
+
+  // End() should be called once the asynchronous task has finished.
+  // Every call to Begin() should eventually result in a call to End().
+  void End();
+
+private:
+  void QueueTick();
+  wgpu::Device const device_;
+  uint64_t count_ = 0;
+  bool tick_queued_ = false;
+  TimeoutId::Generator timeoutIdGenerator;
+};
+
+// AsyncTask is a RAII helper for calling AsyncRunner::Begin() on construction,
+// and AsyncRunner::End() on destruction.
+class AsyncTask {
+public:
+  inline AsyncTask(AsyncTask &&) = default;
+
+  // Constructor.
+  // Calls AsyncRunner::Begin()
+  explicit inline AsyncTask(std::shared_ptr<AsyncRunner> runner)
+      : runner_(std::move(runner)) {
+    runner_->Begin();
+  }
+
+  // Destructor.
+  // Calls AsyncRunner::End()
+  inline ~AsyncTask() { runner_->End(); }
+
+private:
+  AsyncTask(const AsyncTask &) = delete;
+  AsyncTask &operator=(const AsyncTask &) = delete;
+  std::shared_ptr<AsyncRunner> runner_;
```

Comment:
We generally don't use `std::shared_ptr` in the codebase. Instead, we use `kj::Refcounted`. Is there a specific reason you chose `std::shared_ptr` here for this?

Reply:
The specific reason is that I knew the standard `std::shared_ptr`, and hadn't heard about `kj::Refcounted`. :) I changed it, but is there some advantage?